deepmd_utils/model_format/output_def.py

```diff
@@ -0,0 +1,278 @@
+          xxxx_derv_c, xxxx_derv_r are reserved names that should
+          not be used to define variables.
+    shape
+          The shape of the variable. e.g. energy should be [1],
+          dipole should be [3], polarizabilty should be [3,3].
+    reduciable
+          If the variable is reduced.
+    differentiable
+          If the variable is differentiated with respect to coordinates
+          of atoms and cell tensor (pbc case). Only reduciable variable
+          are differentiable.
+
+    """
+
+    def __init__(
+        self,
+        name: str,
+        shape: Union[List[int], Tuple[int]],
+        reduciable: bool = False,
+        differentiable: bool = False,
+    ):
+        # fitting output must be atomic
+        super().__init__(name, shape, atomic=True)
+        self.reduciable = reduciable
+        self.differentiable = differentiable
+        if not self.reduciable and self.differentiable:
+            raise ValueError("only reduciable variable are differentiable")
+
+
+class FittingOutputDef:
+    """Defines the shapes and other properties of the fitting network outputs.
+
+    It is assume that the fitting network output variables for each
+    local atom. This class defines all the outputs.
+
+    Parameters
+    ----------
+    var_defs
+          List of output variable definitions.
+
+    """
+
+    def __init__(
+        self,
+        var_defs: List[OutputVariableDef] = [],
+    ):
+        self.var_defs = {vv.name: vv for vv in var_defs}
+
+    def __getitem__(
+        self,
+        key,
+    ) -> OutputVariableDef:
+        return self.var_defs[key]
+
+    def get_data(self) -> Dict[str, OutputVariableDef]:
+        return self.var_defs
+
+    def keys(self):
+        return self.var_defs.keys()
+
+
+class ModelOutputDef:
+    """Defines the shapes and other properties of the model outputs.
+
+    The model reduce and differentiate fitting outputs if applicable.
+    If a variable is named by foo, then the reduced variable is called
+    foo_redu, the derivative w.r.t. coordinates is called foo_derv_r
+    and the derivative w.r.t. cell is called foo_derv_c.
+
+    Parameters
+    ----------
+    fit_defs
+          Definition for the fitting net output
+
+    """
+
+    def __init__(
+        self,
+        fit_defs: FittingOutputDef,
+    ):
+        self.def_outp = fit_defs
+        self.def_redu = do_reduce(self.def_outp)
+        self.def_derv_r, self.def_derv_c = do_derivative(self.def_outp)
+        self.var_defs = {}
+        for ii in [
+            self.def_outp.get_data(),
+            self.def_redu,
+            self.def_derv_c,
+            self.def_derv_r,
+        ]:
+            self.var_defs.update(ii)
+
+    def __getitem__(self, key) -> VariableDef:
+        return self.var_defs[key]
+
+    def get_data(self, key) -> Dict[str, VariableDef]:
+        return self.var_defs
+
+    def keys(self):
+        return self.var_defs.keys()
+
```

Comment:
**Suggestion:** The `VariableDef`, `OutputVariableDef`, `FittingOutputDef`, and `ModelOutputDef` classes have similar `__init__` methods. Consider using inheritance or composition to reduce code duplication and improve maintainability. [maintainability]
```suggestion
class VariableDef:
    def __init__(
        self,
        name: str,
        shape: Union[List[int], Tuple[int]],
        atomic: bool = True,
    ):
        self.name = name
        self.shape = list(shape)
        self.atomic = atomic

class OutputVariableDef(VariableDef):
    def __init__(
        self,
        name: str,
        shape: Union[List[int], Tuple[int]],
        reduciable: bool = False,
        differentiable: bool = False,
    ):
        super().__init__(name, shape, atomic=True)
        self.reduciable = reduciable
        self.differentiable = differentiable
        if not self.reduciable and self.differentiable:
            raise ValueError("only reduciable variable are differentiable")

class FittingOutputDef(VariableDef):
    def __init__(
        self,
        var_defs: List[OutputVariableDef] = [],
    ):
        super().__init__(name, shape, atomic=True)
        self.var_defs = {vv.name: vv for vv in var_defs}

class ModelOutputDef(FittingOutputDef):
    def __init__(
        self,
        fit_defs: FittingOutputDef,
    ):
        super().__init__(name, shape, atomic=True)
        self.def_outp = fit_defs
        self.def_redu = do_reduce(self.def_outp)
        self.def_derv_r, self.def_derv_c = do_derivative(self.def_outp)
        self.var_defs = {}
        for ii in [
            self.def_outp.get_data(),
            self.def_redu,
            self.def_derv_c,
            self.def_derv_r,
        ]:
            self.var_defs.update(ii)
```